cmd/run.go

```diff
@@ -105,17 +113,95 @@ func (c *cmdRun) run(cmd *cobra.Command, args []string) (err error) {
 		return err
 	}
 
-	// TODO: create a MetricsEngine here and add its ingester to the list of
-	// outputs (unless both NoThresholds and NoSummary were enabled)
+	executionState := execScheduler.GetState()
+	metricsEngine, err := engine.NewMetricsEngine(executionState)
+	if err != nil {
+		return err
+	}
+	if !testRunState.RuntimeOptions.NoSummary.Bool || !testRunState.RuntimeOptions.NoThresholds.Bool {
+		// We'll need to pipe metrics to the MetricsEngine if either the
+		// thresholds or the end-of-test summary are enabled.
+		outputs = append(outputs, metricsEngine.CreateIngester())
+	}
+
+	if !testRunState.RuntimeOptions.NoSummary.Bool {
+		defer func() {
+			logger.Debug("Generating the end-of-test summary...")
+			summaryResult, hsErr := test.initRunner.HandleSummary(globalCtx, &lib.Summary{
+				Metrics:         metricsEngine.ObservedMetrics,
+				RootGroup:       testRunState.Runner.GetDefaultGroup(),
+				TestRunDuration: executionState.GetCurrentTestRunDuration(),
+				NoColor:         c.gs.flags.noColor,
+				UIState: lib.UIState{
+					IsStdOutTTY: c.gs.stdOut.isTTY,
+					IsStdErrTTY: c.gs.stdErr.isTTY,
+				},
+			})
+			if hsErr == nil {
+				hsErr = handleSummaryResult(c.gs.fs, c.gs.stdOut, c.gs.stdErr, summaryResult)
+			}
+			if hsErr != nil {
+				logger.WithError(hsErr).Error("failed to handle the end-of-test summary")
+			}
+		}()
+	}
 
-	// TODO: remove this completely
-	// Create the engine.
-	initBar.Modify(pb.WithConstProgress(0, "Init engine"))
-	engine, err := core.NewEngine(testRunState, execScheduler, outputs)
+	// Create and start the outputs. We do it quite early to get any output URLs
+	// or other details below. It also allows us to ensure when they have
+	// flushed their samples and when they have stopped in the defer statements.
+	initBar.Modify(pb.WithConstProgress(0, "Starting outputs"))
+	outputManager := output.NewManager(outputs, logger, func(err error) {
+		if err != nil {
+			logger.WithError(err).Error("Received error to stop from output")
+		}
+		// TODO: attach run status and exit code?
+		runAbort(err)
+	})
+	samples := make(chan metrics.SampleContainer, test.derivedConfig.MetricSamplesBufferSize.Int64)
+	waitOutputsFlushed, stopOutputs, err := outputManager.Start(samples)
 	if err != nil {
 		return err
 	}
-	engine.AbortFn = runSubAbort
+	defer func() {
+		logger.Debug("Stopping outputs...")
+		// We call waitOutputsFlushed() below because the threshold calculations
+		// need all of the metrics to be sent to the MetricsEngine before we can
+		// calculate them one last time. We need the threshold calculated here,
+		// since they may change the run status for the outputs.
```

Comment:
Instead, to do this by the output manager, could we do an explicit call to the ingester? I think it would simplify a bit the logic.

Reply:
hmm, maybe, not sure it would be simpler though :thinking: here in the `defer` statements, we just have the following sequence after the test run finishes:
1. close the `samples` channel
2. wait for all outputs to finish flushing their metrics
3. finalize the thresholds (unless `--no-thresholds` was used)
4. stop all outputs

to me, this makes sense even if we completely ignore thresholds :thinking: and I am not sure how we can simplify it, when the piping of the metrics also happens by the `output.Manager` :thinking:

Reply:
I think it could be helpful to add these 4 points to the comment so we have one central place where we can read the flow without going across the entire code of the Run method. I think it could be helpful when we will work on #2430.

Reply:
:+1:, but let's add this in the PR that converts the remaining tests, I'll merge this one shortly